File: SudokuKMP/shared/src/commonMain/kotlin/com/nextjedi/sudokustreak/engine/SudokuValidator.kt
package com.nextjedi.sudokustreak.engine

import com.nextjedi.sudokustreak.model.SudokuGrid

class SudokuValidator {

    fun isValidMove(grid: Array<IntArray>, row: Int, col: Int, num: Int): Boolean {
        // Check row
        for (x in 0 until 9) {
            if (grid[row][x] == num) return false
        }
        // Check column
        for (x in 0 until 9) {
            if (grid[x][col] == num) return false
        }
        // Check 3x3 box
        val startRow = (row / 3) * 3
        val startCol = (col / 3) * 3
        for (i in 0 until 3) {
            for (j in 0 until 3) {
                if (grid[startRow + i][startCol + j] == num) return false
            }
        }
        return true
    }

    fun getCandidates(grid: Array<IntArray>, row: Int, col: Int): List<Int> =
        (1..9).filter { isValidMove(grid, row, col, it) }

    fun isComplete(grid: SudokuGrid): Boolean =
        grid.all { row -> row.all { cell -> cell.value != 0 } }

    fun isValidSolution(grid: SudokuGrid): Boolean {
        val nums = (1..9).toSet()
        // Check rows
        for (row in grid) {
            if (row.map { it.value }.toSet() != nums) return false
        }
        // Check columns
        for (col in 0 until 9) {
            if ((0 until 9).map { grid[it][col].value }.toSet() != nums) return false
        }
        // Check 3x3 boxes
        for (boxRow in 0 until 3) {
            for (boxCol in 0 until 3) {
                val values = mutableSetOf<Int>()
                for (i in 0 until 3) {
                    for (j in 0 until 3) {
                        values.add(grid[boxRow * 3 + i][boxCol * 3 + j].value)
                    }
                }
                if (values != nums) return false
            }
        }
        return true
    }
}

File: SudokuKMP/iosApp/iosApp/Engine/SudokuValidator.swift
import Foundation

class SudokuValidator {

    func isValidMove(grid: [[Int]], row: Int, col: Int, num: Int) -> Bool {
        // Check row
        for c in 0..<9 where grid[row][c] == num {
            return false
        }
        // Check column
        for r in 0..<9 where grid[r][col] == num {
            return false
        }
        // Check 3x3 box
        let boxRow = (row / 3) * 3
        let boxCol = (col / 3) * 3
        for r in boxRow..<(boxRow + 3) {
            for c in boxCol..<(boxCol + 3) where grid[r][c] == num {
                return false
            }
        }
        return true
    }

    func getCandidates(grid: [[Int]], row: Int, col: Int) -> [Int] {
        guard grid[row][col] == 0 else { return [] }
        return (1...9).filter { isValidMove(grid: grid, row: row, col: col, num: $0) }
    }

    func isComplete(grid: SudokuGrid) -> Bool {
        grid.allSatisfy { row in row.allSatisfy { $0.value != 0 } }
    }

    func isValidSolution(grid: SudokuGrid) -> Bool {
        let intGrid = toIntGrid(grid)
        // Check all rows have 1-9
        for r in 0..<9 {
            if Set(intGrid[r]) != Set(1...9) { return false }
        }
        // Check all columns have 1-9
        for c in 0..<9 {
            let col = (0..<9).map { intGrid[$0][c] }
            if Set(col) != Set(1...9) { return false }
        }
        // Check all 3x3 boxes have 1-9
        for br in stride(from: 0, to: 9, by: 3) {
            for bc in stride(from: 0, to: 9, by: 3) {
                var box: [Int] = []
                for r in br..<(br + 3) {
                    for c in bc..<(bc + 3) {
                        box.append(intGrid[r][c])
                    }
                }
                if Set(box) != Set(1...9) { return false }
            }
        }
        return true
    }
}
